pygradle-plugin/src/main/groovy/com/linkedin/gradle/python/tasks/ParallelWheelGenerationTask.java

```diff
@@ -0,0 +1,166 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.gradle.python.tasks;
+
+import com.linkedin.gradle.python.PythonExtension;
+import com.linkedin.gradle.python.extension.PlatformTag;
+import com.linkedin.gradle.python.extension.PythonDetails;
+import com.linkedin.gradle.python.extension.PythonTag;
+import com.linkedin.gradle.python.util.PackageInfo;
+import com.linkedin.gradle.python.util.internal.TaskTimer;
+import com.linkedin.gradle.python.wheel.WheelCache;
+import org.apache.commons.io.FileUtils;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.api.tasks.InputFiles;
+import org.gradle.api.tasks.Internal;
+import org.gradle.api.tasks.OutputDirectory;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.internal.logging.progress.ProgressLogger;
+import org.gradle.internal.logging.progress.ProgressLoggerFactory;
+import org.gradle.process.ExecResult;
+
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class ParallelWheelGenerationTask extends DefaultTask {
+
+    private static final Logger logger = Logging.getLogger(ParallelWheelGenerationTask.class);
+
+    private FileCollection filesToConvert;
+    private File cacheDir;
+    private PythonExtension extension;
+    private PythonDetails pythonDetails;
+    private AtomicInteger counter = new AtomicInteger();
+
+    @TaskAction
+    public void buildWheels() {
+
+        ProgressLoggerFactory progressLoggerFactory = getServices().get(ProgressLoggerFactory.class);
+        ProgressLogger progressLogger = progressLoggerFactory.newOperation(ParallelWheelGenerationTask.class);
+        progressLogger.setDescription("Building wheels");
+
+        PythonTag pythonTag = PythonTag.findTag(getProject(), getPythonDetails());
+        PlatformTag platformTag = PlatformTag.makePlatformTag(getProject(), getPythonDetails());
+
+        WheelCache wheelCache = new WheelCache(cacheDir, pythonTag, platformTag);
+
+        progressLogger.started();
+
+        TaskTimer taskTimer = new TaskTimer();
+
+        Set<File> files = getFilesToConvert().getFiles();
+        int totalSize = files.size();
+        files.parallelStream().forEach(file -> {
+            PackageInfo packageInfo = PackageInfo.fromPath(file.getPath());
```

Comment:
How do I exclude certain dependency from this generic wheel building.
Let's assume I have several projects checked out.
For all but one of them I need a wheel for package xxx. However, for one of them I need a custom built xxx with different options or environment.
The existing wheel in the cache would not be satisfactory.
I need a way to skip using a generic wheel here and rebuild from sdist for that project
I don't currently see a way to do that with this approach.

Reply:
You don't. We would be a generic on first, then build a different once second. If the wheel can't be built, it will fail, and ignore silently. The pip-install will then fallback to doing it the old way.

Reply:
Can you be a little more specific?
Here's an example. I want to install the package `Pillow` but pass to it option `--disable-tiff`. The generic will with tiff enabled will not cut it for me.

Can I prevent the build of generic Pilllow wheel at all?

If not, can I at least get an override build with my option, assuming this generic build happens before the regular dependency install and my regular dependency install takes care to ignore the generic wheel?

Reply:
At the moment you can't. When your in-progress change lands (package specific options) they would be excluded from the wheel generation and from being picked up in the wheel options.